clients/src/main/java/org/apache/kafka/clients/consumer/internals/CoordinatorRequestManager.java

```diff
@@ -0,0 +1,289 @@
+                new FindCoordinatorRequestHandler(),
+                null);
+    }
+
+    /**
+     * Mark the current coordinator null and return the old coordinator. Return an empty Optional
+     * if the current coordinator is unknown.
+     * @param cause why the coordinator is marked unknown
+     * @return Optional coordinator node that can be null.
+     */
+    protected void markCoordinatorUnknown(final String cause, final long currentTimeMs) {
+        if (this.coordinator != null) {
+            log.info("Group coordinator {} is unavailable or invalid due to cause: {}. "
+                            + "Rediscovery will be attempted.", this.coordinator, cause);
+            this.coordinator = null;
+            timeMarkedUnknownMs = currentTimeMs;
+        } else {
+            long durationOfOngoingDisconnect = Math.max(0, currentTimeMs - timeMarkedUnknownMs);
+            if (durationOfOngoingDisconnect > this.rebalanceTimeoutMs)
+                log.debug("Consumer has been disconnected from the group coordinator for {}ms",
+                        durationOfOngoingDisconnect);
+        }
+    }
+
+    private void onSuccessfulResponse(
+            final FindCoordinatorResponseData.Coordinator coordinator,
+            final long currentTimeMS) {
+        // use MAX_VALUE - node.id as the coordinator id to allow separate connections
+        // for the coordinator in the underlying network client layer
+        int coordinatorConnectionId = Integer.MAX_VALUE - coordinator.nodeId();
+
+        this.coordinator = new Node(
+                coordinatorConnectionId,
+                coordinator.host(),
+                coordinator.port());
+        log.info("Discovered group coordinator {}", coordinator);
+        coordinatorRequestState.reset();
+        return;
+    }
+
+    private void handleFailedCoordinatorResponse(
+            final Exception exception,
+            final long currentTimeMs) {
+        coordinatorRequestState.updateLastFailedAttempt(currentTimeMs);
+        markCoordinatorUnknown("coordinator unavailable", currentTimeMs);
+
+        if (exception == Errors.GROUP_AUTHORIZATION_FAILED.exception()) {
+            log.debug("FindCoordinator request failed due to authorization error {}", exception.getMessage());
+            errorHandler.handle(GroupAuthorizationException.forGroupId(this.groupId));
+            return;
+        }
+
+        if (exception instanceof RetriableException) {
+            log.debug("FindCoordinator request failed due to retriable exception", exception);
+            return;
+        }
+
+        log.warn("FindCoordinator request failed due to fatal exception", exception);
+        errorHandler.handle(exception);
+    }
+
+    public void onResponse(final FindCoordinatorResponse response, final long currentTimeMs, final Throwable t) {
+        // handle Runtime exception
+        if (t != null) {
+            log.error("FindCoordinator request failed due to {}", t.getMessage());
```

Comment:
Why don't we call `handleFailedCoordinatorResponse` here? Also, we should get in the habit of including the full stack trace in error messages.